server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapStarTreeBuilder.java

```diff
@@ -66,34 +62,136 @@ public List<StarTreeDocument> getStarTreeDocuments() {
+    public Long getDimensionValue(int docId, int dimensionId) {
         return starTreeDocuments.get(docId).dimensions[dimensionId];
     }
 
+    /**
+     * Sorts and aggregates all the documents of the segment based on dimension and metrics configuration
+     *
+     * @param dimensionReaders List of docValues readers to read dimensions from the segment
+     * @param metricReaders List of docValues readers to read metrics from the segment
+     * @return Iterator of star-tree documents
+     *
+     */
     @Override
-    public Iterator<StarTreeDocument> sortAndAggregateStarTreeDocuments() throws IOException {
-        int numDocs = totalSegmentDocs;
-        StarTreeDocument[] starTreeDocuments = new StarTreeDocument[numDocs];
-        for (int currentDocId = 0; currentDocId < numDocs; currentDocId++) {
-            starTreeDocuments[currentDocId] = getSegmentStarTreeDocument(currentDocId);
+    public Iterator<StarTreeDocument> sortAndAggregateSegmentDocuments(
+        SequentialDocValuesIterator[] dimensionReaders,
+        List<SequentialDocValuesIterator> metricReaders
+    ) throws IOException {
+        StarTreeDocument[] starTreeDocuments = new StarTreeDocument[totalSegmentDocs];
+        for (int currentDocId = 0; currentDocId < totalSegmentDocs; currentDocId++) {
+            // TODO : fast exit if all dimensions are null ( indicating all iterators are exhausted )
+            // TODO : if all the dimensions are null, then we can skip adding the document
+            // TODO : if we come out of this loop with the very first document dimensions coming with doc iterators exhausted,
+            // then we need to return empty star tree
+            // TODO : we can save empty iterator for dimensions which are not part of segment
+            starTreeDocuments[currentDocId] = getSegmentStarTreeDocument(currentDocId, dimensionReaders, metricReaders);
         }
-
         return sortAndAggregateStarTreeDocuments(starTreeDocuments);
     }
 
+    @Override
+    public void build(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        build(mergeStarTrees(starTreeValuesSubs));
+    }
+
+    /**
+     * Sorts and aggregates the star-tree documents from multiple segments and builds star tree based on the newly
+     * aggregated star-tree documents
+     *
+     * @param starTreeValuesSubs StarTreeValues from multiple segments
+     * @return iterator of star tree documents
+     */
+    Iterator<StarTreeDocument> mergeStarTrees(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        return sortAndAggregateStarTreeDocuments(mergeStarTreeValues(starTreeValuesSubs), true);
+    }
+
+    /**
+     * Returns an array of all the starTreeDocuments from all the segments
+     *
+     * @param starTreeValuesSubs StarTreeValues from multiple segments
+     * @return array of star tree documents
+     */
+    StarTreeDocument[] mergeStarTreeValues(List<StarTreeValues> starTreeValuesSubs) throws IOException {
```

Comment:
More than merge, this is like a flatten across list<list>. Can we rename accordingly?